File: code/3.无重复字符的最长子串.java
import java.util.HashMap;
import java.util.Map;

/*
 * @lc app=leetcode.cn id=3 lang=java
 *
 * [3] 无重复字符的最长子串
 */

// @lc code=start
class Solution {
    public int lengthOfLongestSubstring(String s) {
        // 滑动窗口
        int l = 0, r = 0;
        int max = 0;
        Map<Character,Integer> lastIndex = new HashMap<>();
        while (l < s.length() && r < s.length()) {
            if(lastIndex.containsKey(s.charAt(r))){
                l = Math.max(lastIndex.get(s.charAt(r))+1, l);
            }
            lastIndex.put(s.charAt(r), r);
            max = Math.max(max,r-l+1);
            r++;
        }
        return max;
    }
}
// @lc code=end

File: code/3.无重复字符的最长子串.c
/*
 * @lc app=leetcode.cn id=3 lang=c
 *
 * [3] 无重复字符的最长子串
 */

// @lc code=start


int lengthOfLongestSubstring(char * s){
    int dict[128];
    memset(dict, -1, sizeof(dict));
    int res = 0, l = 0, r = 0;
    while(r<strlen(s)){
        if(dict[s[r]]+1 > l){
            l = dict[s[r]]+1;
        }
        dict[s[r]] = r;
        if(res<r-l+1) res = r-l+1;
        ++r;
    }
    return res;
}
// @lc code=end

